examples/dnn_introduction3_ex.cpp

```diff
@@ -0,0 +1,170 @@
+// The contents of this file are in the public domain. See LICENSE_FOR_EXAMPLE_PROGRAMS.txt
+/*
+    This is an example illustrating the use of the deep learning tools from the
+    dlib C++ Library.  I'm assuming you have already read the dnn_introduction_ex.cpp and
+    the dnn_introduction2_ex.cpp examples.  So in this example program I'm going to go
+    over a transfer learning example, which includes:
+        - Defining a layer visitor to modify the some network parameters for fine-tuning
+        - Using pretrained layers of a network for another task
+*/
+
+#include <dlib/dnn.h>
+#include <iostream>
+
+// This header file includes a generic definition of the most common ResNet architectures
+#include "resnet.h"
+
+using namespace std;
+using namespace dlib;
+
+// In this simple example we will show how to load a pretrained network and use it for a
+// different task.  In particular, we will load a ResNet50 trained on ImageNet, adjust
+// some of its parameters and use it as a pretrained backbone for some metric learning
+// task.
+
+// Let's start by defining a network that will use the ResNet50 backbone from resnet.h
+namespace model
+{
+    template<template<typename> class BN>
+    using net_type = loss_metric<
+        fc_no_bias<128,
+        avg_pool_everything<
+        typename resnet<BN>::template backbone_50<
+        input_rgb_image
+        >>>>;
+
+    using train = net_type<bn_con>;
+    using infer = net_type<affine>;
+}
+
+// Next, we define a layer visitor that will modify the weight decay of a network.  The
+// main interest of this class is to show how one can define custom visitors that modify
+// some network parameters.
+class visitor_weight_decay_multiplier
+{
+public:
+
+    visitor_weight_decay_multiplier(double new_weight_decay_multiplier_) :
+        new_weight_decay_multiplier(new_weight_decay_multiplier_) {}
+
+    template <typename T>
+    void set_new_weight_decay_mulitplier(T& l) const
+    {
+        set_weight_decay_multiplier(l, new_weight_decay_multiplier);
+        set_bias_weight_decay_multiplier(l, new_weight_decay_multiplier);
```

Comment:
Same thing here.  The effective bias weight decay is the overall weight decay times the specific bias weight decay.  So only call the first method.